python/lancedb/embeddings/functions.py

```diff
@@ -0,0 +1,224 @@
+#  Copyright (c) 2023. LanceDB Developers
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import json
+from abc import ABC, abstractmethod
+from typing import List, Optional, Union
+
+import numpy as np
+import pyarrow as pa
+from cachetools import cached
+from pydantic import BaseModel
+
+
+class EmbeddingFunctionRegistry:
+    """
+    This is a singleton class used to register embedding functions
+    and fetch them by name. It also handles serializing and deserializing
+    """
+
+    @classmethod
+    def get_instance(cls):
+        return REGISTRY
+
+    def __init__(self):
+        self._functions = {}
+
+    def register(self):
+        """
+        This creates a decorator that can be used to register
+        an EmbeddingFunctionModel.
+        """
+
+        # This is a decorator for a class that inherits from BaseModel
+        # It adds the class to the registry
+        def decorator(cls):
+            if not issubclass(cls, EmbeddingFunctionModel):
+                raise TypeError("Must be a subclass of EmbeddingFunctionModel")
+            if cls.__name__ in self._functions:
+                raise KeyError(f"{cls.__name__} was already registered")
+            self._functions[cls.__name__] = cls
+            return cls
+
+        return decorator
+
+    def reset(self):
+        """
+        Reset the registry to its initial state
+        """
+        self._functions = {}
+
+    def load(self, name: str):
+        """
+        Fetch an embedding function class by name
+        """
+        return self._functions[name]
+
+    def parse_functions(self, metadata: Optional[dict]) -> dict:
+        """
+        Parse the metadata from an arrow table and
+        return a mapping of the vector column to the
```

Comment:
So the key is input column? or the generated embedding (output) column?

Reply:
The key is the vector column.
This is so we can have fast lookup when executing a search query